features/terrain.py

```diff
@@ -0,0 +1,252 @@
+from lettuce import world, before, after
+import os.path
+import psycopg2
+import requests
+import subprocess
+import shutil
+import tempfile
+import time
+import yaml
+
+
+class PatroniController(object):
+    PATRONI_CONFIG = '{}.yml'
+    """ starts and stops individual patronis"""
+
+    def __init__(self):
+        self._output_dir = None
+        self._patroni_path = None
+        self._connections = {}
+        self._config = {}
+        self._connstring = {}
+        self._cursors = {}
+        self._log = {}
+        self._processes = {}
+
+    @property
+    def patroni_path(self):
+        if self._patroni_path is None:
+            cwd = os.path.realpath(__file__)
+            while True:
+                path, entry = os.path.split(cwd)
+                cwd = path
+                if entry == 'features' or cwd == '/':
+                    break
+            self._patroni_path = cwd
+        return self._patroni_path
+
+    def start(self, pg_name, max_wait_limit=15):
+        if not self._is_running(pg_name):
+            if pg_name in self._processes:
+                del self._processes[pg_name]
+            cwd = self.patroni_path
+            self._log[pg_name] = open(os.path.join(self._output_dir, 'patroni_{0}.log'.format(pg_name)), 'a')
+
+            self._config[pg_name] = self._make_patroni_test_config(pg_name)
+
+            p = subprocess.Popen(['python', 'patroni.py', self._config[pg_name]],
+                                 stdout=self._log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
+            if not (p and p.pid and p.poll() is None):
+                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
+            self._processes[pg_name] = p
+        # wait while patroni is available for queries, but not more than 10 seconds.
+        for tick in range(max_wait_limit):
+            if self.query(pg_name, "SELECT 1", fail_ok=True) is not None:
+                break
+            time.sleep(1)
+        else:
+            assert False,\
+                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+
+    def stop(self, pg_name, kill=False):
+        while self._is_running(pg_name):
+            if not kill:
+                self._processes[pg_name].terminate()
+            else:
+                self._processes[pg_name].kill()
+            time.sleep(1)
+        self._log.get('pg_name') and self._log[pg_name].close()
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2115598839/issues/source?bid=3127594&fileBranchId=3144623#l68)